kotlinx-coroutines-core/common/src/flow/internal/AbstractSharedFlow.kt

```diff
@@ -10,19 +10,66 @@ import kotlin.jvm.*
 @JvmField
 internal val EMPTY_RESUMES = arrayOfNulls<Continuation<Unit>?>(0)
 
+/**
+ * A slot allocated to a collector when it subscribes to a shared flow and freed when the collector unsubscribes.
+ */
 internal abstract class AbstractSharedFlowSlot<F> {
+    /**
+     * Try marking this slot as allocated for the given [flow]. Only call this under the [flow]'s lock.
+     *
+     * Returns `false` if the slot is already allocated to some other collector.
+     */
     abstract fun allocateLocked(flow: F): Boolean
-    abstract fun freeLocked(flow: F): Array<Continuation<Unit>?> // returns continuations to resume after lock
+
+    /**
+     * Mark this slot as available for reuse. Only call this under the [flow]'s lock.
+     *
+     * Returns an array of continuations that need to be resumed after the lock is released.
+     * These continuations represent suspended emitters that were waiting for the slowest collector to move on
+     * so that the next value can be placed into the buffer.
+     */
+    abstract fun freeLocked(flow: F): Array<Continuation<Unit>?>
 }
 
-internal abstract class AbstractSharedFlow<S : AbstractSharedFlowSlot<*>> : SynchronizedObject() {
-    protected var slots: Array<S?>? = null // allocated when needed
+/**
+ * A common data structure for `StateFlow` and `SharedFlow`.
+ */
+internal abstract class AbstractSharedFlow<This, S : AbstractSharedFlowSlot<This>> : SynchronizedObject() {
```

Comment:
Is `This` a new Kotlin feature to refer to the type of the instance's subclass?

Reply:
It's only a generic parameter name. Could as well be named T. That said, I was also startled when I saw `This`. Possibly rename into something less misleading.